src/main/kotlin/com/team4099/robot2023/subsystems/elevator/ElevatorIONEO.kt

```diff
@@ -0,0 +1,188 @@
+
+        leaderSparkMax.openLoopRampRate = ElevatorConstants.RAMP_RATE.inPercentOutputPerSecond
+        followerSparkMax.openLoopRampRate = ElevatorConstants.RAMP_RATE.inPercentOutputPerSecond
+
+        leaderSparkMax.idleMode = CANSparkMax.IdleMode.kBrake
+        followerSparkMax.idleMode = CANSparkMax.IdleMode.kBrake
+
+        // makes follower motor output exact same power as leader
+        followerSparkMax.follow(leaderSparkMax, ElevatorConstants.FOLLOWER_INVERTED)
+
+        leaderPIDController.ff = 0.0
+
+        leaderSparkMax.burnFlash()
+        followerSparkMax.burnFlash()
+
+        MotorChecker.add(
+            "Elevator",
+            "Extension",
+            MotorCollection(
+                mutableListOf(
+                    Neo(leaderSparkMax, "Leader Extension Motor"),
+                    Neo(followerSparkMax, "Follower Extension Motor")
+                ),
+                ElevatorConstants.LEADER_STATOR_CURRENT_LIMIT,
+                30.celsius,
+                ElevatorConstants.LEADER_STATOR_CURRENT_LIMIT - 30.amps,
+                90.celsius
+            ),
+        )
+    }
+
+    override fun updateInputs(inputs: ElevatorIO.ElevatorInputs) {
+        inputs.elevatorPosition = leaderSensor.position
+
+        inputs.elevatorVelocity = leaderSensor.velocity
+
+        // voltage in * percent out
+        inputs.leaderAppliedVoltage = leaderSparkMax.busVoltage.volts * leaderSparkMax.appliedOutput
+
+        inputs.leaderStatorCurrent = leaderSparkMax.outputCurrent.amps
+
+        // BatteryVoltage * SupplyCurrent = percentOutput * BatteryVoltage * StatorCurrent
+        // AppliedVoltage = percentOutput * BatteryVoltage
+        // SupplyCurrent = (percentOutput * BatteryVoltage / BatteryVoltage) * StatorCurrent =
+        // percentOutput * statorCurrent
+
+        inputs.leaderSupplyCurrent =
+            inputs.leaderStatorCurrent * leaderSparkMax.appliedOutput.absoluteValue
+
+        inputs.leaderTempCelcius = leaderSparkMax.motorTemperature.celsius
+
+        // voltage in * percent out
+        inputs.followerAppliedVoltage = leaderSparkMax.busVoltage.volts * followerSparkMax.appliedOutput
+
+        inputs.followerStatorCurrent = followerSparkMax.outputCurrent.amps
+
+        inputs.followerSupplyCurrent =
+            inputs.followerStatorCurrent * followerSparkMax.appliedOutput.absoluteValue
+
+        inputs.followerTempCelcius = followerSparkMax.motorTemperature.celsius
+
+        inputs.leaderRawPosition = leaderSparkMax.encoder.position
+
+        inputs.followerRawPosition = followerSparkMax.encoder.position
+
+        Logger.recordOutput("Elevator/leaderRawRotations", leaderSparkMax.encoder.position)
+    }
+
+    /**
+     * Sets the voltage of the elevator motors but also checks to make sure elevator doesn't exceed
+     * limit
+     *
+     * @param voltage the voltage to set the motor to
+     */
+    override fun setOutputVoltage(voltage: ElectricalPotential) {
```

Comment:
should wrap this in an if statement that checks whether our elevator is either
1. `<0.5 inches` and applying a negative voltage
2. `> max_height-0.5 inches` and applying a positive voltage

if either of these are true, we don't want to be able to run the motor bc that will break elevator

Reply:
did not realize this logic was in `Elevator.kt` should live in docstrings of this function, though. io.setVoltage() does not handle any soft limit checks